[AArch64]: Refactor target parser to use Bitset.

hassnaaHamdi opened the pull request:
Use Bitset instead of BitMasking for the Architecture Extensions, because the number of extensions will exceed the bitmask max size.

hassnaaHamdi committed:
[AArch64]: Refactor target parser to use BitVector.

Use BitVector instead of BitMasking for the Architecture Extensions,
because the number of extensions will exceed the bitmask max size.

Change-Id: Ib4dcf7ee7a400d93ce89c02895ebf4cec2da0f7f

DavidSpickett commented:
With just the bitvector change this is fine.

I don't disagree with splitting the test files but it should be in its own commit/PR.

sdesmalen-arm commented on llvm/unittests/TargetParser/AArch64TargetParserTest.cpp:
Would it have been possible to change ARMCPUTestParams to give it two constructors, one taking a uint64_t mask (for the ARM features), and the other constructor taking a BitVector (for the AArch64 features), and then to have the classes work on a BitVector under the hood?

For example, it's possible to convert a bitmask -> BitVector with a simple loop:
```uint64_t Mask = ... // the mask you want to convert to a BitVector
BitVector V(64);
for (unsigned I=0; I<64; ++I) {
  if ((Mask >> I) & 1)
    V.set(I);
}
```

sdesmalen-arm reviewed (commented):
Seems like a sensible refactor, we're going to run out of bits eventually.

pratlucas commented on llvm/include/llvm/TargetParser/AArch64TargetParser.h:
```diff
@@ -96,64 +97,65 @@ static_assert(FEAT_MAX <= 64,
-  AEK_RASv2 =       1ULL << 54, // FEAT_RASv2
-  AEK_ITE =         1ULL << 55, // FEAT_ITE
-  AEK_GCS =         1ULL << 56, // FEAT_GCS
+enum ArchExtKind : int {
```
Since the enum will no longer represent a bit map, can it be made an `enum class`?

hassnaaHamdi replied:
I still need to use the enum member as integers, because the members will be used as indexes.

pratlucas commented on llvm/include/llvm/TargetParser/AArch64TargetParser.h:
```diff
@@ -315,23 +317,23 @@ struct ArchInfo {
-inline constexpr ArchInfo ARMV9_2A  = { VersionTuple{9, 2}, AProfile, "armv9.2-a", "+v9.2a", (ARMV9_1A.DefaultExts)};
-inline constexpr ArchInfo ARMV9_3A  = { VersionTuple{9, 3}, AProfile, "armv9.3-a", "+v9.3a", (ARMV9_2A.DefaultExts | AArch64::AEK_MOPS | AArch64::AEK_HBC)};
-inline constexpr ArchInfo ARMV9_4A  = { VersionTuple{9, 4}, AProfile, "armv9.4-a", "+v9.4a", (ARMV9_3A.DefaultExts | AArch64::AEK_SPECRES2 | AArch64::AEK_CSSC | AArch64::AEK_RASv2)};
+inline ArchInfo ARMV8A    = { VersionTuple{8, 0}, AProfile, "armv8-a", "+v8a", (BitVector(AEK_EXTENTIONS_NUM).set(AArch64::AEK_FP).set(AArch64::AEK_SIMD)), };
```
Could we have a constructor for `ArchInfo` that takes a list of `ArchExtKind`s as an argument (e.g. an initialiser_list)? This would make these lines less verbose and a bit more readable.

hassnaaHamdi replied:
Hi,
Thanks for looking at my patch,
Now I'm using Bitset which supports that.

topperc commented:
Does this end up creating a global constructor?

Could llvm::Bitset be used instead. It has a constexpr constructor.

hassnaaHamdi committed:
[AArch64]: Refactor target parser to use Bitset.

Use Bitset instead of BitMasking for the Architecture Extensions,
because the number of extensions will exceed the bitmask max size.

Change-Id: Id869744514ce89b9ecb50bac033b45eda4d8ceac

hassnaaHamdi commented:
Hi,
Thanks for your comments.
I think that I have resolved them.
Now I'm using Bitset instead of Bitvector, as it has constexpr constructor.
Also I dropped splitting the TargetParserTest, and now It's a single test as previous.

topperc commented on llvm/unittests/TargetParser/TargetParserTest.cpp:
```diff
@@ -69,6 +70,24 @@ std::string FormatExtensionFlags(uint64_t Flags) {
 
+std::string SerializeExtensionFlags(Bitset<AArch64::AEK_EXTENTIONS_MAX> Flags) {
+  std::string SerializedFlags;
+  int HexaValue = 0;
```
Hexa -> Hex.

topperc commented on llvm/unittests/TargetParser/TargetParserTest.cpp:
```diff
@@ -69,6 +70,24 @@ std::string FormatExtensionFlags(uint64_t Flags) {
+  for(unsigned int i = 0; i < AArch64::AEK_EXTENTIONS_MAX; i ++) {
+    HexaValue <<= 1;
+    HexaValue |= (int)Flags[i];
+    if ((i+1)%4 == 0) {
```
This assumes AArch64::AEK_EXTENTIONS_MAX is a multiple of 4 so there is no remainder after the loop. Probably should static_assert that.

topperc commented on llvm/include/llvm/TargetParser/AArch64TargetParser.h:
```diff
@@ -96,64 +97,65 @@ static_assert(FEAT_MAX <= 64,
+  AEK_RASv2 =         55, // FEAT_RASv2
+  AEK_ITE =           56, // FEAT_ITE
+  AEK_GCS =           57, // FEAT_GCS
+  AEK_EXTENTIONS_MAX =  64
```
EXTENTIONS -> EXTENSIONS

hassnaaHamdi committed:
Code Format

Change-Id: I746c95d9cf0006044e89be88447a866462df0515

hassnaaHamdi committed:
assert MAX extensions is divisible by 4 to correctly calc hex value of enabled bits.
Fix type.

Change-Id: I23e125c93c82a8fa6fb77b30468ddd5bf7d9c13b

hassnaaHamdi committed:
fix missing replacement of AEK_EXTENTIONS_MAX

Change-Id: I2494d0d4e59723eba7a206c7615834d522f86a16

hassnaaHamdi commented on llvm/unittests/TargetParser/TargetParserTest.cpp:
```diff
@@ -1605,13 +1691,14 @@ TEST(TargetParserTest, AArch64ExtensionFeatures) {
+  Bitset<AArch64::AEK_EXTENSIONS_MAX> ExtVal;
   for (auto Ext : Extensions)
-    ExtVal |= Ext;
+    ExtVal.set(Ext);
```
It's works only for the object. Ext should be an object of Bitset to work.

hassnaaHamdi committed:
Resolve review comments.

- discard formating that is not related to my changes.
- improve SerializeExtensionFlags() function.
- use ArchExtKind::AEK_EXTENSIONS_LAST instead of AEK_EXTENSIONS_MAX,
  to represent the size of the extensions, given that it's 1-based.

Change-Id: Id3d453ed8cf5a9b3fb170892b8a64793acad4031

hassnaaHamdi commented:
Comments are resolved.

hassnaaHamdi commented:
Hi @sdesmalen-arm 
Do you have further comments ?

hassnaaHamdi committed:
Resolve review comments.

- Use type alias declaration for Bitset.
- Use function overloading instead of function template,
  to avoid error prone code.

Change-Id: I277c32b10d75e36e3c522ca6f736f23374bc87d6

hassnaaHamdi commented on llvm/unittests/TargetParser/TargetParserTest.cpp:
```diff
@@ -18,38 +18,39 @@
     // AEK_NONE is not meant to be shown to the user so the target parser
     // does not recognise it. It is relevant here though.
-    if (Flags & AArch64::AEK_NONE)
+    if (Flags.test(AArch64::AEK_NONE))
```
Thanks for the explanation :))

sdesmalen-arm commented on llvm/include/llvm/TargetParser/AArch64TargetParser.h:
```diff
@@ -318,23 +319,23 @@ struct ArchInfo {
-inline constexpr ArchInfo ARMV8_2A  = { VersionTuple{8, 2}, AProfile, "armv8.2-a", "+v8.2a", (ARMV8_1A.DefaultExts | Bitset<AArch64::AEK_EXTENSIONS_LAST>({AArch64::AEK_RAS}))};
-inline constexpr ArchInfo ARMV8_3A  = { VersionTuple{8, 3}, AProfile, "armv8.3-a", "+v8.3a", (ARMV8_2A.DefaultExts | Bitset<AArch64::AEK_EXTENSIONS_LAST>({AArch64::AEK_RCPC}))};
-inline constexpr ArchInfo ARMV8_4A  = { VersionTuple{8, 4}, AProfile, "armv8.4-a", "+v8.4a", (ARMV8_3A.DefaultExts | Bitset<AArch64::AEK_EXTENSIONS_LAST>({AArch64::AEK_DOTPROD}))};
+inline constexpr ArchInfo ARMV8A    = { VersionTuple{8, 0}, AProfile, "armv8-a", "+v8a", (AArch64::ExtensionBitset({AArch64::AEK_FP, AArch64::AEK_SIMD})), };
```
nit: can you move the `(AArch64::ExtensionBitset(...))` to a new line?

sdesmalen-arm reviewed (approved):
LGTM

hassnaaHamdi committed:
Resolve review comments

- code format.

Change-Id: I5ca736f312a29f86781f29e333f12dcbe9610015

hassnaaHamdi commented:
Hi @hctim 
Thanks for notifying me :) I see that someone else fixed the issue.

topperc commented on llvm/include/llvm/TargetParser/AArch64TargetParser.h:
```diff
@@ -96,65 +97,67 @@ static_assert(FEAT_MAX <= 64,
-  AEK_ITE =         1ULL << 55, // FEAT_ITE
-  AEK_GCS =         1ULL << 56, // FEAT_GCS
+enum ArchExtKind : unsigned {
+  AEK_NONE =          1,
```
Why is bit 0 skipped?
